File: tests/docker-compat/python-client/tests/lists.py
"""Category 4: List Operations tests."""

from .framework import (
    register, run_test, key, assert_equal, assert_true,
    assert_equal_int, assert_list_equal,
)


def test_lists(r, prefix):
    results = []
    cat = "lists"

    def t_push():
        k = key(prefix, "list:push")
        try:
            r.lpush(k, "a")
            r.rpush(k, "b")
            r.lpush(k, "c", "d")   # d, c, a, b
            r.rpush(k, "e", "f")   # d, c, a, b, e, f
            vals = r.lrange(k, 0, -1)
            assert_list_equal(["d", "c", "a", "b", "e", "f"], vals)
        finally:
            r.delete(k)
    results.append(run_test("LPUSH_RPUSH_single_and_multiple", cat, t_push))

    def t_pop():
        k = key(prefix, "list:pop")
        try:
            r.rpush(k, "a", "b", "c")
            left = r.lpop(k)
            assert_equal("a", left)
            right = r.rpop(k)
            assert_equal("c", right)
        finally:
            r.delete(k)
    results.append(run_test("LPOP_RPOP", cat, t_pop))

    def t_lrange():
        k = key(prefix, "list:range")
        try:
            r.rpush(k, "a", "b", "c", "d", "e")
            vals = r.lrange(k, 1, 3)
            assert_list_equal(["b", "c", "d"], vals)
        finally:
            r.delete(k)
    results.append(run_test("LRANGE", cat, t_lrange))

    def t_llen():
        k = key(prefix, "list:len")
        try:
            r.rpush(k, "a", "b", "c")
            assert_equal_int(3, r.llen(k))
        finally:
            r.delete(k)
    results.append(run_test("LLEN", cat, t_llen))

    def t_lindex():
        k = key(prefix, "list:idx")
        try:
            r.rpush(k, "a", "b", "c")
            assert_equal("b", r.lindex(k, 1))
            assert_equal("c", r.lindex(k, -1))
        finally:
            r.delete(k)
    results.append(run_test("LINDEX", cat, t_lindex))

    def t_lset():
        k = key(prefix, "list:set")
        try:
            r.rpush(k, "a", "b", "c")
            r.lset(k, 1, "B")
            assert_equal("B", r.lindex(k, 1))
        finally:
            r.delete(k)
    results.append(run_test("LSET", cat, t_lset))

    def t_linsert():
        k = key(prefix, "list:ins")
        try:
            r.rpush(k, "a", "c")
            r.linsert(k, "BEFORE", "c", "b")  # a, b, c
            r.linsert(k, "AFTER", "c", "d")   # a, b, c, d
            vals = r.lrange(k, 0, -1)
            assert_list_equal(["a", "b", "c", "d"], vals)
        finally:
            r.delete(k)
    results.append(run_test("LINSERT_BEFORE_AFTER", cat, t_linsert))

    def t_lrem():
        k = key(prefix, "list:rem")
        try:
            r.rpush(k, "a", "b", "a", "c", "a")
            n = r.lrem(k, 2, "a")
            assert_equal_int(2, n)
            vals = r.lrange(k, 0, -1)
            assert_list_equal(["b", "c", "a"], vals)
        finally:
            r.delete(k)
    results.append(run_test("LREM", cat, t_lrem))

    def t_ltrim():
        k = key(prefix, "list:trim")
        try:
            r.rpush(k, "a", "b", "c", "d", "e")
            r.ltrim(k, 1, 3)
            vals = r.lrange(k, 0, -1)
            assert_list_equal(["b", "c", "d"], vals)
        finally:
            r.delete(k)
    results.append(run_test("LTRIM", cat, t_ltrim))

    def t_lpos():
        k = key(prefix, "list:pos")
        try:
            r.rpush(k, "a", "b", "c", "b", "d")
            pos = r.lpos(k, "b")
            assert_equal_int(1, pos)
        finally:
            r.delete(k)
    results.append(run_test("LPOS", cat, t_lpos))

    def t_lmpop():
        k = key(prefix, "list:mpop")
        try:
            r.rpush(k, "a", "b", "c", "d")
            result = r.lmpop(1, k, direction="LEFT", count=2)
            assert_true(result is not None, "LMPOP should return a result")
            popped_key, vals = result
            assert_equal(k, popped_key)
            assert_list_equal(["a", "b"], vals)
        finally:
            r.delete(k)
    results.append(run_test("LMPOP", cat, t_lmpop))

    return results


register("lists", test_lists)

File: tests/docker-compat/js-client/tests/lists.js
const { register, runTest, key, assertEqual, assertEqualInt, assertArrayEqual } = require('./framework');

register('lists', async (redis, prefix) => {
  const results = [];
  const cat = 'lists';

  results.push(await runTest('LPUSH_RPUSH_single_and_multiple', cat, async () => {
    const k = key(prefix, 'list:push');
    await redis.lpush(k, 'a');
    await redis.rpush(k, 'b');
    await redis.lpush(k, 'c', 'd');  // d, c, a, b
    await redis.rpush(k, 'e', 'f');  // d, c, a, b, e, f
    const vals = await redis.lrange(k, 0, -1);
    assertArrayEqual(['d', 'c', 'a', 'b', 'e', 'f'], vals);
    await redis.del(k);
  }));

  results.push(await runTest('LPOP_RPOP', cat, async () => {
    const k = key(prefix, 'list:pop');
    await redis.rpush(k, 'a', 'b', 'c');
    const left = await redis.lpop(k);
    assertEqual('a', left);
    const right = await redis.rpop(k);
    assertEqual('c', right);
    await redis.del(k);
  }));

  results.push(await runTest('LRANGE', cat, async () => {
    const k = key(prefix, 'list:range');
    await redis.rpush(k, 'a', 'b', 'c', 'd', 'e');
    const vals = await redis.lrange(k, 1, 3);
    assertArrayEqual(['b', 'c', 'd'], vals);
    await redis.del(k);
  }));

  results.push(await runTest('LLEN', cat, async () => {
    const k = key(prefix, 'list:len');
    await redis.rpush(k, 'a', 'b', 'c');
    const n = await redis.llen(k);
    assertEqualInt(3, n);
    await redis.del(k);
  }));

  results.push(await runTest('LINDEX', cat, async () => {
    const k = key(prefix, 'list:idx');
    await redis.rpush(k, 'a', 'b', 'c');
    const val1 = await redis.lindex(k, 1);
    assertEqual('b', val1);
    const val2 = await redis.lindex(k, -1);
    assertEqual('c', val2);
    await redis.del(k);
  }));

  results.push(await runTest('LSET', cat, async () => {
    const k = key(prefix, 'list:set');
    await redis.rpush(k, 'a', 'b', 'c');
    await redis.lset(k, 1, 'B');
    const val = await redis.lindex(k, 1);
    assertEqual('B', val);
    await redis.del(k);
  }));

  results.push(await runTest('LINSERT_BEFORE_AFTER', cat, async () => {
    const k = key(prefix, 'list:ins');
    await redis.rpush(k, 'a', 'c');
    await redis.linsert(k, 'BEFORE', 'c', 'b');  // a, b, c
    await redis.linsert(k, 'AFTER', 'c', 'd');   // a, b, c, d
    const vals = await redis.lrange(k, 0, -1);
    assertArrayEqual(['a', 'b', 'c', 'd'], vals);
    await redis.del(k);
  }));

  results.push(await runTest('LREM', cat, async () => {
    const k = key(prefix, 'list:rem');
    await redis.rpush(k, 'a', 'b', 'a', 'c', 'a');
    const n = await redis.lrem(k, 2, 'a');
    assertEqualInt(2, n);
    const vals = await redis.lrange(k, 0, -1);
    assertArrayEqual(['b', 'c', 'a'], vals);
    await redis.del(k);
  }));

  results.push(await runTest('LTRIM', cat, async () => {
    const k = key(prefix, 'list:trim');
    await redis.rpush(k, 'a', 'b', 'c', 'd', 'e');
    await redis.ltrim(k, 1, 3);
    const vals = await redis.lrange(k, 0, -1);
    assertArrayEqual(['b', 'c', 'd'], vals);
    await redis.del(k);
  }));

  results.push(await runTest('LPOS', cat, async () => {
    const k = key(prefix, 'list:pos');
    await redis.rpush(k, 'a', 'b', 'c', 'b', 'd');
    const pos = await redis.call('LPOS', k, 'b');
    assertEqualInt(1, pos);
    await redis.del(k);
  }));

  results.push(await runTest('LMPOP', cat, async () => {
    const k = key(prefix, 'list:mpop');
    await redis.rpush(k, 'a', 'b', 'c', 'd');
    const result = await redis.call('LMPOP', 1, k, 'LEFT', 'COUNT', 2);
    assertEqual(k, result[0]);
    assertArrayEqual(['a', 'b'], result[1]);
    await redis.del(k);
  }));

  return results;
});
